mdb/src/usb.rs

```diff
@@ -0,0 +1,153 @@
+            _ => None,
+        }
+    }
+}
+
+pub fn usb_devices() -> Result<UsbDeviceList> {
+    let context = GlobalContext::default();
+    //context.set_log_level(LogLevel::Debug);
+    Ok(UsbDeviceList(context.devices()?))
+}
+
+pub struct UsbDeviceList(DeviceList<GlobalContext>);
+
+impl UsbDeviceList {
+    pub fn iter(&self) -> UsbDevices {
+        UsbDevices(self.0.iter())
+    }
+}
+
+pub struct UsbDevices<'a>(Devices<'a, GlobalContext>);
+
+impl<'a> Iterator for UsbDevices<'a> {
+    type Item = Result<UsbDevice>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        while let Some(device) = self.0.next() {
+            if let Some(res) = UsbDevice::new(device).transpose() {
+                return Some(res);
+            }
+        }
+        None
+    }
+}
+
+#[derive(Debug, Eq, PartialEq)]
+pub struct UsbDevice {
+    handle: DeviceHandle<GlobalContext>,
+    serial: String,
+    protocol: Protocol,
+    config: u8,
+    iface: u8,
+    setting: u8,
+    ep_read: u8,
+    ep_write: u8,
+}
+
+impl UsbDevice {
+    fn new(device: Device<GlobalContext>) -> Result<Option<Self>> {
+        let device_desc = device.device_descriptor()?;
+        let config_desc = device.active_config_descriptor()?;
+        for iface in config_desc.interfaces() {
+            for iface_desc in iface.descriptors() {
+                if let Some(protocol) = Protocol::new(&iface_desc) {
+                    let ep_read = iface_desc
+                        .endpoint_descriptors()
+                        .filter(|ep| ep.transfer_type() == TransferType::Bulk)
+                        .filter(|ep| ep.direction() == Direction::In)
+                        .map(|ep| ep.address())
+                        .next()
+                        .ok_or_else(|| anyhow::anyhow!("invalid endpoint"))?;
+                    let ep_write = iface_desc
+                        .endpoint_descriptors()
+                        .filter(|ep| ep.transfer_type() == TransferType::Bulk)
+                        .filter(|ep| ep.direction() == Direction::Out)
+                        .map(|ep| ep.address())
+                        .next()
+                        .ok_or_else(|| anyhow::anyhow!("invalid endpoint"))?;
+                    let handle = device.open().map_err(error)?;
+                    let serial = handle.read_serial_number_string_ascii(&device_desc)?;
+                    return Ok(Some(Self {
+                        handle,
+                        serial,
+                        protocol,
+                        config: config_desc.number(),
+                        iface: iface_desc.interface_number(),
+                        setting: iface_desc.setting_number(),
+                        ep_read,
+                        ep_write,
+                    }));
+                }
+            }
+        }
+        Ok(None)
+    }
+
+    pub fn open(serial: &str, protocol: Protocol) -> Result<Self> {
+        let mut device = usb_devices()?
```

Comment:
We'll have to refactor this API somewhat so that we aren't iterating USB devices twice. At the same time this means devices from `usb_devices()` are not opened, but will be opened only when you call `UsbDevice::open()` and get the same type back.